staging/src/k8s.io/apiserver/pkg/server/filters/with_early_late_annotations.go

```diff
@@ -0,0 +1,133 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+	utilsclock "k8s.io/utils/clock"
+)
+
+type lifecycleEvent interface {
+	// Name returns the name of the signal, useful for logging.
+	Name() string
+
+	// Signaled returns a channel that is closed when the underlying event
+	// has been signaled. Successive calls to Signaled return the same value.
+	Signaled() <-chan struct{}
+
+	// SignaledAt returns the time the event was signaled. If SignaledAt is
+	// invoked before the event is signaled (zero, false) will be returned.
+	SignaledAt() (time.Time, bool)
+}
+
+// WithShutdownLateAnnotation, if added to the handler chain, tracks the
+// incoming request(s) after the apiserver has initiated the graceful
+// shutdown, and annoates the audit event for these request(s) with
+// diagnostic information.
+// This enables us to identify the actor(s)/load balancer(s) that are sending
+// requests to the apiserver late during the server termination.
+func WithShutdownLateAnnotation(handler http.Handler, shutdownInitiated lifecycleEvent, delayDuration time.Duration) http.Handler {
+	return withShutdownLateAnnotation(handler, shutdownInitiated, delayDuration, utilsclock.RealClock{})
+}
+
+func withShutdownLateAnnotation(handler http.Handler, shutdownInitiated lifecycleEvent, delayDuration time.Duration, clock utilsclock.PassiveClock) http.Handler {
+	if shutdownInitiated == nil || clock == nil || delayDuration == time.Duration(0) {
+		klog.Warning("WithShutdownLateAnnotation filter is not being added to the handler chain")
+		return handler
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		select {
+		case <-shutdownInitiated.Signaled():
+		default:
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		requestor, exists := request.UserFrom(req.Context())
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+
+		defer handler.ServeHTTP(w, req)
+
+		var loopback bool
+		if requestor.GetName() == user.APIServerUser {
+			loopback = true
+		}
+
+		initiatedAt, _ := shutdownInitiated.SignaledAt()
+		elapsedSince := clock.Since(initiatedAt)
+		var veryLate bool
+		// TODO: 80% is the threshold, if requests arrive after 80% of
+		// shutdown-delay-duration elapses we annotate the request as late=true.
+		if float64(elapsedSince)/float64(delayDuration) > 0.80 {
+			veryLate = true
+		}
+
+		value := fmt.Sprintf("since=%s elapsed=%s shutdown-delay-duration=%s loopback=%t late=%t",
```

Comment:
rather than use late=true, a different annotation is easier to explain.

Reply:
if the first annotation is parsed, then do we see value in adding another annotation?